strawberryfields/engine.py

```diff
@@ -73,50 +73,56 @@
 
 """
 
+import abc
 from collections.abc import Sequence
 
 from .backends import load_backend
 from .backends.base import (NotApplicableError, BaseBackend)
 
 
-class Engine:
-    r"""Quantum program executor engine.
+class Result:
+    """Result of a quantum computation.
 
-    :meth:`run` executes :class:`.Program` instances on the chosen backend, and makes
-    measurement results available via the :class:`.RegRef` instances.
+    Represents the results of the execution of a quantum program.
+    Returned by :meth:`.BaseEngine.run`.
+    """
+    def __init__(self, samples):
+        #: BaseState: quantum state object returned by a local backend, if any
+        self.state = None
+        #: List[List[Number]]: measurement samples, shape == (modes, shots)
+        self.samples = samples
+
+    def __str__(self):
+        
+        return 'Result: {} subsystems, state: {}\n samples: {}'.format(len(self.samples), self.state, self.samples)
+
+
+class BaseEngine(abc.ABC):
+    r"""ABC for quantum program executor engines.
 
     Args:
-        backend (str): name of the backend
-        host    (str): URL of the remote backend host, or None if a local backend is used
-    Keyword Args:
+        backend (str): backend short name
+        backend_options (Dict[str, Any]): keyword arguments for the backend
     """
-    def __init__(self, backend, host=None, **kwargs):
-        #: list[Program]: list of Programs that have been run
+    def __init__(self, backend, backend_options={}):
+        #: str: short name of the backend
+        self.backend_name = backend
+        #: Dict[str, Any]: keyword arguments for the backend
+        self.backend_options = backend_options.copy()  # dict is mutable
+        #: List[Program]: list of Programs that have been run
         self.run_progs = []
-        #: str: URL of the remote backend host
-        self.host = host
-        #: dict[str]: keyword arguments for the backend
-        self.kwargs = kwargs
-
-        if isinstance(backend, str):
-            #: str: short name of the backend
-            self.backend_name = backend
-            #: BaseBackend: backend for executing the quantum circuit
-            self.backend = load_backend(backend)
-        elif isinstance(backend, BaseBackend):
-            self.backend_name = backend._short_name
-            self.backend = backend
-        else:
-            raise TypeError('backend must be a string or a BaseBackend instance.')
+        #: List[List[Number]]: latest measurement results, shape == (modes, shots)
+        self.samples = None
```

Comment:
The engine can execute several Program instances in succession, and after each one it stores the measurement results in `self.samples`. The engine needs to do this because the same Program instance could be next executed by a different engine (possibly with a different Program history), changing the measurement results stored in its RegRefs.

Reply:
ok that makes sense for RegRefs, but won't there also be a `Results` class (possibly under a different name) returned by an engine which also stores this information? Isn't it redundant to have the engine also store it? 

I feel like there's some special use-case involving `RegRefTransform`s here, but I couldn't figure it out

Reply:
Each engine instance represents a single quantum computation, which IMO makes it the primary place to store that information. If a Program contains RRTs (gates that depend on previously measured values) the engine needs to be able to access those samples before it can execute the Program.

OTOH `Result.samples` is copy of `BaseEngine.samples`. The `Result` class is a bit redundant in the sense that all that information could be accessed manually through the engine instance, but I think it's good to have it from a usability perspective. One can think of a Result instance as a snapshot of the state of the computation at one point in time.